Analysis/bin/ShuffleMergeSpectral.cxx

```diff
@@ -0,0 +1,681 @@
+    run::Argument<unsigned> n_threads{"n-threads", "number of threads", 1};
+    run::Argument<unsigned> seed{"seed", "random seed to initialize the generator used for sampling", 1234567};
+    run::Argument<std::string> disabled_branches{"disabled-branches",
+                                                 "list of branches to disabled in the input tuples", ""};
+    run::Argument<std::string> path_spectrum{"input_spec", "input path with spectrums for all the samples"};
+    run::Argument<std::string> tau_ratio{"tau_ratio", "ratio of tau types in the final spectrum"};
+};
+
+namespace {
+struct SourceDesc {
+    using Tau = tau_tuple::Tau;
+    using TauTuple = tau_tuple::TauTuple;
+    using SampleType = analysis::SampleType;
+
+    SourceDesc(const std::string& _name,const std::string&  _group_name, const std::vector<std::string>& _file_names,
+               double _weight, const std::set<std::string>& _disabled_branches) :
+        name(_name), group_name(_group_name), file_names(_file_names), disabled_branches(_disabled_branches), weight(_weight),
+        current_n_processed(0), files_n_total(_file_names.size()), entries_file(666), total_n_processed(0)
+    {
+        if(file_names.empty())
+            throw analysis::exception("Empty list of files for the source '%1%'.") % name;
+        if(weight <= 0)
+            throw analysis::exception("Invalid source weight for the source '%1%'.") % name;
+        if(!files_n_total)
+            throw analysis::exception("Empty source '%1%'.") % name;
+
+    }
+
+    SourceDesc(const SourceDesc&) = delete;
+    SourceDesc& operator=(const SourceDesc&) = delete;
+
+    bool DoNextStep()
+    {
+
+      if(current_file_index == files_n_total-1 && current_n_processed >= entries_file) return false;
+
+      while(!current_file_index || current_n_processed == entries_file) {
+          if(!current_file_index)
+              current_file_index = 0;
+          else
+              ++(*current_file_index);
+          if(*current_file_index >= file_names.size())
+              throw analysis::exception("The expected number of events = %1% is bigger than the actual number of"
+                                        " events in source '%2%'.") % entries_file % name;
+          current_n_processed = 0;
+          const std::string& file_name = file_names.at(*current_file_index);
+          current_tuple.reset();
+          current_file = root_ext::OpenRootFile(file_name);
+          current_tuple = std::make_shared<TauTuple>("taus", current_file.get(), true, disabled_branches);
+          entries_file = current_tuple->GetEntries();
+          if(!entries_file)
+            throw analysis::exception("Root file inside %1% is empty.") % name;
+      }
+      ++total_n_processed;
+      current_tuple->GetEntry(current_n_processed++);
+      (*current_tuple)().dataset_id = std::hash<std::string>{}(name);
+      (*current_tuple)().dataset_group_id = std::hash<std::string>{}(group_name);
```

Comment:
- computing hash for each entry could be computationally costly. Can be done in the constructor
- std::hash does not guarantee reproducibility between different machines/compilers/environments. The table of hash<->name correspondence should be stored in an auxiliary TTree or in a csv/json file.